### PR TITLE
Lammps Dump Parser

### LINKED ISSUE #3843: LAMMPS dump file unwrap trajectory and import velocities and forces
## Is your feature request related to a problem? ##
LAMMPS trajectories can be written in two way, (1) xu yu zu in dump to allow diffusion coefficients to be computed. and (2) x y z ix iy iz, where the image flags keep track of which image of the wrapped box that a particle is in. In theory, this means one could unwrap the box, however this information is lost upon import into MDAnalysis.
In addition, MDAnalysis does not import velocities and forces from dump files.

## Additional context ##
I'm aware of the other LAMMPS related pull request but all changes there refer to the LAMMPSParser in topology, not MDAnalysis/coordinates/LAMMPS.py

### PR DESCRIPTION
**Pull request by jaclark5**
Fixes #3843 #3741 

Changes made in this Pull Request:
 - Allow box translation upon importing LAMMPS dump file
 - Allows coordinates to be unwrapped with dump file image flags
 - Allows import of velocities and forces from lammps dump file

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [X] Issue raised/referenced?

### CODE REVIEW

File: package/MDAnalysis/coordinates/LAMMPS.py
Comment (richardjgowers) [inline package/MDAnalysis/coordinates/LAMMPS.py]: This looks like it could scale badly.  If we convert `coords` to a numpy array, we can check its shape and work wholly in numpy... something like:
```python
coords = np.array(coords)

if coords.shape[1] == 6:  # also check it's either 3 or 6, else we've got something weird
    images = coords[:, 3:]
    coords = coords[:, :3]
    coords += images * ts.dimensions[:3]
```

File: package/MDAnalysis/coordinates/LAMMPS.py
Comment (hmacdope) [inline package/MDAnalysis/coordinates/LAMMPS.py]: The order here is important, because these are ordered in terms of precedence. As imageflag is probably quite rare (??) I would possibly put it last?

File: package/MDAnalysis/coordinates/LAMMPS.py
Comment (hmacdope) [inline package/MDAnalysis/coordinates/LAMMPS.py]: Same I would put this last for consistency

File: package/MDAnalysis/coordinates/LAMMPS.py
Comment (hmacdope) [inline package/MDAnalysis/coordinates/LAMMPS.py]: You can construct the `ts` and then set `has_velocities` or `has_forces` attr on it later. See my next comment. :)

File: package/MDAnalysis/coordinates/LAMMPS.py
Comment (hmacdope) [inline package/MDAnalysis/coordinates/LAMMPS.py]: Rather than reopening the file I think this check should happen where the columns are parsed for the ATOM card, on line 633 and below.

File: package/MDAnalysis/coordinates/LAMMPS.py
Comment (hmacdope) [inline package/MDAnalysis/coordinates/LAMMPS.py]: It would be great if you could make this kwarg more specific, something like `unwrap_imageflag` as unfortunately there is a lot of  name collision with the word `unwrap` in MDAnalysis.

You will also need some logic to raise an exception if the option is used without `imageflag_unscaled` coords

File: package/MDAnalysis/coordinates/LAMMPS.py
Comment (hmacdope) [inline package/MDAnalysis/coordinates/LAMMPS.py]: Nitpick but `np.float32` is currently canonical for coords.

File: package/MDAnalysis/coordinates/LAMMPS.py
Comment (hmacdope) [inline package/MDAnalysis/coordinates/LAMMPS.py]: Perhaps `NotImplementedError`?

File: package/MDAnalysis/coordinates/LAMMPS.py
Comment (hmacdope) [inline package/MDAnalysis/coordinates/LAMMPS.py]: Curious as to why this is necessary. Have we missed something about LAMMPS boxes? 🤔

**Review by hmacdope — CHANGES_REQUESTED** [review]
Thanks for the contribution @jaclark5! It's looking great so far. See my comments. We will also need some tests with a small `imageflag` containing testfile. :)

It may also be better to split the `imageflag` and forces+velocities stuff into separate PRs to avoid this one becoming super crowded but up to you at the end of the day.

File: package/MDAnalysis/coordinates/LAMMPS.py
Comment (hmacdope) [inline package/MDAnalysis/coordinates/LAMMPS.py]:
```suggestion
    .. versionchanged:: 2.4.0
```

File: package/MDAnalysis/coordinates/LAMMPS.py
Comment (hmacdope) [inline package/MDAnalysis/coordinates/LAMMPS.py]: Please set a default value for the kwarg in `__init__` eg `unwrap_images=Flase` and then change this check.

File: package/MDAnalysis/coordinates/LAMMPS.py
Comment (hmacdope) [inline package/MDAnalysis/coordinates/LAMMPS.py]: No need for explicit line continuation here. Just use the implicit continuation of
```python
( "blah blah"
  "blah ....")
```

**Review by hmacdope — CHANGES_REQUESTED** [review]
Some more quick comments from a read over.

File: testsuite/MDAnalysisTests/datafiles.py
Comment (jbarnoud) [inline testsuite/MDAnalysisTests/datafiles.py:513]: These two files need to be referenced in `__all__` at the top of the file. This is why the tests fail.

File: package/CHANGELOG
Comment (hmacdope) [inline package/CHANGELOG]: Doubled from above.

File: package/MDAnalysis/coordinates/LAMMPS.py
Comment (hmacdope) [inline package/MDAnalysis/coordinates/LAMMPS.py]: Is there a need for this `#NoteHere`?

File: package/MDAnalysis/coordinates/LAMMPS.py
Comment (hmacdope) [inline package/MDAnalysis/coordinates/LAMMPS.py]: This should cover all of the `vx, vy, vz` attributes rather than just vx

File: package/MDAnalysis/coordinates/LAMMPS.py
Comment (hmacdope) [inline package/MDAnalysis/coordinates/LAMMPS.py:623]: No need to change this back from an f-string. In general, avoid changing things that are not directly related to your PR or fix as it can mess up git history a bit.

File: package/MDAnalysis/coordinates/LAMMPS.py
Comment (hmacdope) [inline package/MDAnalysis/coordinates/LAMMPS.py]: You will need to add a test that covers this.

**Review by hmacdope — CHANGES_REQUESTED** [review]
Few comments :)

File: package/CHANGELOG
Comment (hmacdope) [inline package/CHANGELOG]:
```suggestion
  *  LammpsDump Reader optionally unwraps trajectories with image flags upon loading (#3843)
```

File: package/MDAnalysis/coordinates/LAMMPS.py
Comment (hmacdope) [inline package/MDAnalysis/coordinates/LAMMPS.py]: Is there a link to the lammps docs on image flags?

File: package/MDAnalysis/coordinates/LAMMPS.py
Comment (hmacdope) [inline package/MDAnalysis/coordinates/LAMMPS.py]: Make this the same check as above with 

```python
all(x in attr_to_col_ix for x in ["fx", "fy", "fz"]
```

File: package/MDAnalysis/coordinates/LAMMPS.py
Comment (hmacdope) [inline package/MDAnalysis/coordinates/LAMMPS.py]: PEP8 indentation needs to match the line above 

```suggestion
                                      "{self.lammps_coordinate_convention} found in timestep")
```

File: package/MDAnalysis/coordinates/LAMMPS.py
Comment (hmacdope) [inline package/MDAnalysis/coordinates/LAMMPS.py]: nitpick but use dtype to make slightly clearer. 

```suggestion
            coords = np.array([fields[dim] for dim in coord_cols], dtype=np.float32)
```

**Review by hmacdope — CHANGES_REQUESTED** [review]
Looking good, just a few suggestions.

File: package/CHANGELOG
Comment (hmacdope) [inline package/CHANGELOG]: Make all the uses of LAMMPSDump the same, `LAMMPSDump`.

currently 3 different versions

File: package/MDAnalysis/coordinates/LAMMPS.py
Comment (hmacdope) [inline package/MDAnalysis/coordinates/LAMMPS.py:710]: @PicoCentauri is this the kind of solution you are after for #3741 ?

**Review by hmacdope — CHANGES_REQUESTED** [review]
Almost there, quick comment and will await some feedback from @PicoCentauri

File: package/MDAnalysis/coordinates/LAMMPS.py
Comment (PicoCentauri) [reply package/MDAnalysis/coordinates/LAMMPS.py:710]: Yes, thanks @jaclark5 this will solve the inconsistency.

**Review by hmacdope — APPROVED** [review]
I'm happy with this now. :) Anyone else want to review or shall we go ahead.

File: testsuite/MDAnalysisTests/coordinates/test_lammps.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/coordinates/test_lammps.py:142]:
```suggestion


```

File: testsuite/MDAnalysisTests/coordinates/test_lammps.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/coordinates/test_lammps.py:129]:
```suggestion


```

File: testsuite/MDAnalysisTests/coordinates/test_lammps.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/coordinates/test_lammps.py:116]:
```suggestion


```

File: package/MDAnalysis/coordinates/LAMMPS.py
Comment (IAlibay) [inline package/MDAnalysis/coordinates/LAMMPS.py]: please revert this, it exceeds 79 characters.

File: package/MDAnalysis/coordinates/LAMMPS.py
Comment (IAlibay) [inline package/MDAnalysis/coordinates/LAMMPS.py]: We really need a `Parameters` section for this parser, this is too much text for folks to read and understand what's going on.

I'm happy having this be addressed in a follow-up PR, but can an issue be raised please?

File: package/MDAnalysis/coordinates/LAMMPS.py
Comment (IAlibay) [inline package/MDAnalysis/coordinates/LAMMPS.py]: Why is this being changed? It doesn't violate pep8 nor does it seem to have any functional changes?

File: package/MDAnalysis/coordinates/LAMMPS.py
Comment (IAlibay) [inline package/MDAnalysis/coordinates/LAMMPS.py]: this is more than 79 chars, please break into two lines

File: package/MDAnalysis/coordinates/LAMMPS.py
Comment (IAlibay) [inline package/MDAnalysis/coordinates/LAMMPS.py]:
```suggestion
        self._unwrap = unwrap_images
```
If the default is `False` then the `if` / `else` shouldn't be necessary?

File: testsuite/MDAnalysisTests/coordinates/test_lammps.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/coordinates/test_lammps.py]: For new code (no need to do this for old stuff here), please use `assert_allclose` here, see: https://github.com/MDAnalysis/mdanalysis/issues/3743

File: testsuite/MDAnalysisTests/coordinates/test_lammps.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/coordinates/test_lammps.py]: As above, please use `assert_allclose` here, see: https://github.com/MDAnalysis/mdanalysis/issues/3743

File: testsuite/MDAnalysisTests/coordinates/test_lammps.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/coordinates/test_lammps.py]: As above, if you're doing a line change, might as well switch to `assert_allcose`.

File: testsuite/MDAnalysisTests/coordinates/test_lammps.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/coordinates/test_lammps.py]: as above

File: testsuite/MDAnalysisTests/coordinates/test_lammps.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/coordinates/test_lammps.py]: as above

File: testsuite/MDAnalysisTests/coordinates/test_lammps.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/coordinates/test_lammps.py]: as above

**Review by IAlibay — CHANGES_REQUESTED** [review]
Couple things on my end, mostly pep8

**Review by IAlibay — APPROVED** [review]
lgtm thanks!

**Comment by jaclark5** [comment]
Hmm I thought this was supposed to merge upon approval, do I need to do anything else?

**Comment by IAlibay** [comment]
> Hmm I thought this was supposed to merge upon approval, do I need to do anything else?

I was just waiting on CI, I'll merge now (we don't have merge on approval set up, it's something the coredevs need to discuss at some point I guess)

**Pull request merged**
